### DISCUSSION TITLE
argocd.argoproj.io/refresh:hard annotation is refreshing the application after every second.

### DISCUSSION THREAD

**Discussion by imuneeeb** [argoproj/argo-cd#19399] [Q&A]
I applied this annotation **_argocd.argoproj.io/refresh:hard_** in template of one of my applicationset and it is refreshing the applications after every single second. I couldnt find the soultion in documentation to limit refreshing. I really apprecite your contribution.

**Comment by crenshaw-dev** [comment]
Adding the annotation to your ApplicationSet `spec.template.metadata.annotations` is a bad way to refresh an app. The AppSet controller will continually apply the annotation to your Application. Instead you should use the UI, CLI, or `kubectl apply` to add the annotation to the apps you want to refresh. The Application controller will refresh the app and then remove the annotation so you don't get stuck in a refresh loop.

**Reply by imuneeeb** [reply]
@crenshaw-dev 
I am using a helm chart and I have to keep changing values file from time to time. It's really difficult to manullay hard refresh the application to pick up new values everytime. Argo doesnt atomatically pickup new changes.
Is there any other automated way to do it?

**Reply by crenshaw-dev** [reply]
Are you writing values into the Application manifest or into git?

If you're writing them into the Application manifest, there should be no need to refresh at all. The application-controller should automatically recognize that the `spec.source.helm.values/valuesObject` field changed and trigger a refresh.

**Reply by imuneeeb** [reply]
@crenshaw-dev 
I am writing them into application manifest `spec.source.helm.valueFiles/path-to-file`. My helm chart creates Role and RoleBinding against the values I add in values file. The application-controller doesnt create new role and rolebinding when I add new values untill i hard refresh it.

**Reply by crenshaw-dev** [reply]
So you're pushing a change to the values file in git? In that case a simple refresh should be enough. And if a refresh isn't fast enough, you can configure git webhooks.

**Reply by imuneeeb** [reply]
@crenshaw-dev 
Yes I am pushing the changes in the git and webhooks refresh the argo on every commit. Actually my valueFiles in helm is overridden by another file from some other directory. So untill I do not hard refresh, argo doesnt reflect new changes.

**Reply by imuneeeb** [reply]
@crenshaw-dev 
Looking forward to listen back from you

**Comment by saikumar13** [comment]
Hi @imuneeeb  
May I ask you to share the app template reference once? I have used the below parameter in the app template, even though my use case is related to helm charts and custom environment-sspecific values. YAML: I have been experiencing some challenges, and the team is complaining about some configmap parameters not reflecting, so I used to do a hard refresh on UI upon request. Later I have added the below parameter. Is this valid? Please suggest.

app-template.yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: my-app
  namespace: argocd
spec:
  destination:
    server: https://kubernetes.default.svc
    namespace: my-namespace
  source:
    repoURL: 'https://github.com/my-repo.git'
    targetRevision: HEAD
    path: my-app-path
  syncPolicy:
    **syncOptions:
      - HardRefresh=true**

Thanks !!

**Reply by imuneeeb** [reply]
@saikumar13 
If you are updating the configmap and values are not reflecting that's a different usecase. Can you share a little bit more detail?

**Reply by saikumar13** [reply]
@imuneeeb  Sure, Sorry for delay in response 

I am running argocd 2.8v  and currently using  argocd app template with multiple sources one of the source pointing to environment specific values yaml file which is BitBucket Repo let say values-dev.yaml, values-qa.yaml etc.,
another source is pointing to chartmusuem with respective chartName and in the charts. we do have values.yaml  and same thing referred in app template.. 
---
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: myapp
  namespace: argo-cd
  finalizers:
  - resources-finalizer.argocd.argoproj.io
spec:
  project: demo-pr  
  sources:
      # This repo and branch will be used to get custom ops values.yaml
    - repoURL: https://mycompany.com/scm/sd/myapp.git
      targetRevision: master
      ref: opsValEnvRepo    
    - repoURL: https://helmrepo.dev.mycmp.com/deploy
      chart: myapp
      targetRevision: 3.2-SNAPSHOT
      helm:
        # Release name override (defaults to application name)
        releaseName: myapp
        valueFiles:
        - values.yaml
        # helm external values
        - $opsValEnvRepo/values-dev.yaml
        - $opsValEnvRepo/imageTags.yaml         
        parameters:
        # namespace parameter will used to override the default metadata.namespace
        - name: "namespace"
          value: default
  destination:
    name: in-cluster
    # The namespace will only be set for namespace-scoped resources that have not set a value for .metadata.namespace
    namespace: {{ index .Values "myapp" "ns" | default "default" }}
  syncPolicy:
    automated: {}
      #prune: {{ .Values.syncPolicy.prune | default false }}
      #selfHeal: {{ .Values.syncPolicy.selfHeal | default false }}
    syncOptions:
      # - ServerSideApply=true
      - CreateNamespace=true
      - HardRefresh=true

This is my app template. 

The issue currently I am facing is that let say in helm section chartRevision is pointing to 3.2-SNAPSHOT 
I have added some parameters in values.yaml  completed CI but chart pushed with same version.. (3.2-SNAPSHOT) as we have changes in values.yaml 
My assumption is that argocd will able to identify the change but it is not until I initiate refresh Hard from UI as I am new to this explored that we  do have annotation with Hard Refresh so if I add argocd.argoproj.io/refresh:hard under annotation will this solve my problem Or any other solution you will recommend. 

Please suggest.

Thanks !!
